app/src/main/java/com/google/android/ground/ui/home/mapcontainer/jobs/JobMapAdapter.kt

```diff
@@ -0,0 +1,398 @@
+      jobCardOpened.value = true
+      selectedFeatureListener(selectedLoi.loi.id)
+    }
+  }
+
+  fun setSelectedFeature(listener: (String?) -> Unit) {
+    selectedFeatureListener = listener
+  }
+
+  fun setCollectDataListener(listener: (DataCollectionEntryPointData) -> Unit) {
+    collectDataListener.value = listener
+  }
+
+  private fun initializeAddLoiButton() {
+    with(basemapLayoutBinding) {
+      (bottomContainer as ViewGroup).addView(
+        ComposeView(bottomContainer.context).apply {
+          setContent {
+            AppTheme {
+              InitializeAddLoiButton {
+                if (newLoiJobs.size == 1) {
+                  // If there's only one job, start data collection on it without showing the
+                  // job modal.
+                  jobModalOpened.value = false
+                  collectDataListener.value(newLoiJobs.first())
+                } else {
+                  jobModalOpened.value = true
+                }
+              }
+              InitializeJobSelectionModal()
+            }
+          }
+        }
+      )
+    }
+  }
+
+  private fun initializeJobCard() {
+    with(basemapLayoutBinding) {
+      (bottomContainer as ViewGroup).addView(
+        ComposeView(bottomContainer.context).apply {
+          setContent { AppTheme { InitializeJobCard() } }
+        }
+      )
+    }
+  }
+
+  private fun closeJobCard() {
+    jobCardOpened.value = false
+    activeLoi.value = null
+    selectedFeatureListener(null)
+  }
+
+  @Composable
+  private fun InitializeAddLoiButton(callback: () -> Unit) {
+    val jobs = remember { newLoiJobs }
+    val jobModalOpened by remember { jobModalOpened }
+    val canUserSubmitData by remember { canUserSubmitData }
+    if (jobs.size == 0 || jobModalOpened || !canUserSubmitData) {
+      return
+    }
+    Row(
+      Modifier.fillMaxWidth(),
+      horizontalArrangement = Arrangement.Center,
+      verticalAlignment = Alignment.CenterVertically,
+    ) {
+      ActionButton(
+        icon = Icons.Filled.Add,
+        contentDescription = stringResource(id = R.string.add_site),
+        callback = callback,
+      )
+    }
+  }
+
+  @Composable
+  private fun InitializeJobSelectionModal() {
+    val jobs = remember { newLoiJobs }
+    var openJobsModal by remember { jobModalOpened }
+    val collectDataCallback by remember { collectDataListener }
+    if (openJobsModal) {
+      basemapLayoutBinding.mapTypeBtn.hide()
+      basemapLayoutBinding.locationLockBtn.hide()
+      menuBinding.hamburgerBtn.hide()
+      Modal(onDismiss = { openJobsModal = false }) {
+        jobs.forEach { job ->
+          JobSelectionRow(job) {
+            collectDataCallback(job)
+            openJobsModal = false
+          }
+          Spacer(Modifier.height(16.dp))
+        }
+      }
+    } else {
+      basemapLayoutBinding.mapTypeBtn.show()
+      basemapLayoutBinding.locationLockBtn.show()
+      menuBinding.hamburgerBtn.show()
```

Comment:
In order to make these class pure composable objects and remove coupling from existing layout elements, can we instead expose states which can then be handled in the caller class (HomescreenMapContainerFragment)?

Something like `onDismissed` or anything else

Reply:
Yes! Done.